cdap-watchdog/src/test/java/co/cask/cdap/logging/save/LogSaverPluginTest.java

```diff
@@ -0,0 +1,476 @@
+                                0, Long.MAX_VALUE, Filter.EMPTY_FILTER, logCallback);
+    Assert.assertEquals(60, logCallback.getEvents().size());
+
+    // Reset checkpoint for metrics plugin
+    resetMetricsPluginCheckPoint();
+
+    // reset the logsaver to start reading from reset offset
+    Set<Integer> partitions = Sets.newHashSet(0, 1);
+    logSaver.unscheduleTasks();
+    logSaver.scheduleTasks(partitions);
+
+    waitTillLogSaverDone(ns1LogBaseDir, "APP_1/flow-FLOW_1/%s", "Test log message 59 arg1 arg2");
+    stopLogSaver();
+
+    //Verify that no more records are processed by LogWriter plugin
+    distributedLogReader.getLog(new FlowletLoggingContext("NS_1", "APP_1", "FLOW_1", "", null, "INSTANCE"),
+                                0, Long.MAX_VALUE, Filter.EMPTY_FILTER, logCallback);
+    Assert.assertEquals(60, logCallback.getEvents().size());
+    // Checkpoint should read 60 for both processor
+    verifyCheckpoint();
+  }
+
+  private void resetMetricsPluginCheckPoint() throws Exception {
+    TypeToken token = new TypeToken<Set<KafkaLogProcessor>> () { };
+    Set<KafkaLogProcessor> processors = (Set<KafkaLogProcessor>) injector.getInstance(Key.get(token.getType(),
+                                                                Names.named(Constants.LogSaver.MESSAGE_PROCESSORS)));
+    for (KafkaLogProcessor processor : processors) {
+      if (processor instanceof  LogMetricsPlugin) {
+        LogMetricsPlugin plugin = (LogMetricsPlugin) processor;
+        CheckpointManager manager = plugin.getCheckPointManager();
+        manager.saveCheckpoint(0, 10);
+        Set<Integer> partitions = Sets.newHashSet(0, 1);
+        plugin.init(partitions);
+      }
+    }
+  }
+
+  @AfterClass
+  public static void shutdown() {
+    txManager.stopAndWait();
+  }
+
+  public void verifyCheckpoint() throws Exception {
+    TypeToken token = new TypeToken<Set<KafkaLogProcessor>> () { };
+    Set<KafkaLogProcessor> processors = (Set<KafkaLogProcessor>) injector.getInstance(Key.get(token.getType(),
+                                                               Names.named (Constants.LogSaver.MESSAGE_PROCESSORS)));
+    for (KafkaLogProcessor processor : processors) {
+      CheckpointManager checkpointManager = getCheckPointManager(processor);
+      Assert.assertEquals(60, checkpointManager.getCheckpoint(0));
+    }
+  }
+
+  private static CheckpointManager getCheckPointManager(KafkaLogProcessor processor) {
+    if (processor instanceof KafkaLogWriterPlugin) {
+      KafkaLogWriterPlugin plugin = (KafkaLogWriterPlugin) processor;
+      return plugin.getCheckPointManager();
+    } else if (processor instanceof LogMetricsPlugin) {
+      LogMetricsPlugin plugin = (LogMetricsPlugin) processor;
+      return plugin.getCheckPointManager();
+    }
+    throw new IllegalArgumentException("Invalid processor");
+  }
+
+  private void testLogRead(LoggingContext loggingContext) throws Exception {
```

Comment:
Yes. I opened a JIRA to fix in 3.0 time frame. https://issues.cask.co/browse/CDAP-2165